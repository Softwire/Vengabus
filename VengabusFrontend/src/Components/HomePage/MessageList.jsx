import React, { Component } from 'react';
import { DataTable } from '../DataTable';
import { MessageBox } from '../MessageBox/MessageBox';
import { truncate } from 'lodash';
import { formatTimeStamp } from '../../Helpers/FormattingHelpers';


export class MessageList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            showMessage: false
        };
    }

    handleMessageClick = (e, row, rowIndex) => {
        const id = row.messageId;
        const body = row.messageBody;
        this.setState({
            showMessage: true,
            messageId: id,
            messageBody: body,
            message: row
        });
    }

    handleClose = () => {
        this.setState({
            showMessage: false
        });
    }

    render() {
        const messageArray = this.props.messageData ? [...this.props.messageData] : this.props.messageData;
        //add a preview of the body to each field which will be the first 30 chars
        const previewLength = 30;
        for (let i = 0; i < (messageArray ? messageArray.length : 0); i++) {
            const currentMessageArray = messageArray[i];
            currentMessageArray.messageBodyPreview = truncate(currentMessageArray.messageBody, { length: previewLength });
            const enqueuedTime = currentMessageArray.predefinedProperties.enqueuedTimeUtc;
            if (enqueuedTime) {
                currentMessageArray.timestamp = formatTimeStamp(enqueuedTime);
            } else {
                currentMessageArray.timestamp = '##-##-#### ##:##:##';
            }
        }


        const colProps = [
            {
                dataField: 'uniqueId',
                hidden: true
            },
            {
                dataField: 'predefinedProperties.messageId',
                text: 'Message Id',
                width: 33,
                headerStyle: this.props.headerStyle,
                search: function (row) {
                    let searchString = row.predefinedProperties.messageId;
                    if (typeof searchString !== "string") {
                        searchString = searchString.toString();
                    }
                    return searchString;
                }
            },
            {
                dataField: 'messageBodyPreview',
                text: 'Message Body',
                width: 33,
                headerStyle: this.props.headerStyle,
                search: true
            },
            {
                dataField: 'timestamp',
                width: 34,
                headerStyle: this.props.headerStyle

            }
        ];

        return (
            <div>
                <DataTable
                    name='MessageTable'
                    colProps={colProps}
                    uniqueKeyColumn='uniqueId'
                    dataToDisplay={messageArray}
                    defaultHover
                    onRowClick={this.handleMessageClick}
                    paginated
                    searchable
                />

                <MessageBox
                    message={this.state.message}
                    show={this.state.showMessage}
                    endpointType={this.props.endpointType}
                    messageType={this.props.messageType}
                    endpointName={this.props.endpointName}
                    endpointParent={this.props.endpointParent}
                    handleClose={this.handleClose}
                    onMessageBoxClose={this.props.refreshMessageTableHandler}
                />
            </div>
        );
    }
}
