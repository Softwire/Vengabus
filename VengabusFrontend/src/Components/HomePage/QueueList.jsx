import React, { Component } from 'react';
import { DataTable } from '../DataTable';
import { formatTimeStamp } from '../../Helpers/FormattingHelpers';
import { palerBlue } from '../../colourScheme';

export class QueueList extends Component {

    render() {
        let queueArray = undefined;
        if (this.props.queueData) {
            queueArray = [...this.props.queueData];

            for (let i = 0; i < queueArray.length; i++) {
                //needs to be cloned
                queueArray[i] = { ...queueArray[i] };
                const currentMessageArray = queueArray[i];
                if (currentMessageArray.mostRecentDeadLetter) {
                    currentMessageArray.mostRecentDeadLetter = formatTimeStamp(currentMessageArray.mostRecentDeadLetter);
                }
            }
        }

        const colProps = [
            {
                dataField: 'name',
                width: 25,
            },
            {
                dataField: 'activeMessageCount',
                width: 25,
                align: 'right'
            },
            {
                dataField: 'deadletterMessageCount',
                width: 25,
                align: 'right'
            },
            {
                dataField: 'mostRecentDeadLetter',
                width: 25,
            }
        ];

        const selectRow = {
            bgColor: palerBlue,
            selected: this.props.currentlySelectedName
        };

        return (
            <DataTable
                name='QueueList'
                colProps={colProps}
                dataToDisplay={queueArray}
                defaultHover
                onRowClick={this.props.clickFunction}
                selectRow={selectRow}
                uniqueKeyColumn='name'
                searchable
            />
        );
    }
}
