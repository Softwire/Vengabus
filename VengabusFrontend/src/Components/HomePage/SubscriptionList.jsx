import React, { Component } from 'react';
import { DataTable } from '../DataTable';
import { formatTimeStamp } from '../../Helpers/FormattingHelpers';
import { palerBlue } from '../../colourScheme';
import { EditSubscriptionButton } from '../Buttons/EditEndpointButton';


export class SubscriptionList extends Component {

    render() {
        const originalSubscriptionProps = this.props.subscriptionData;
        const subscriptionArray = originalSubscriptionProps ? [] : originalSubscriptionProps;
        if (originalSubscriptionProps) {
            for (let i = 0; i < originalSubscriptionProps.length; i++) {
                //needs to be cloned
                subscriptionArray.push({ ...originalSubscriptionProps[i] });
                const currentMessageArray = subscriptionArray[i];
                if (currentMessageArray.mostRecentDeadletter && currentMessageArray.mostRecentDeadletterLoaded) {
                    currentMessageArray.mostRecentDeadletter = formatTimeStamp(currentMessageArray.mostRecentDeadletter);
                }
            }
        }


        const colProps = [
            {
                dataField: 'name',
                width: 23,
                headerStyle: this.props.headerStyle,
                search: true
            },
            {
                //qq this is only a workaround for the table not rerendering itself. Solve this properly later.
                dataField: 'activeMessageCount',
                text: 'Active / Deadletter Message Counts',
                width: 46,
                formatter: (cell, row, rowIndex) => {
                    return (`${row.activeMessageCount} / ${row.deadletterMessageCount}`);
                },
                headerStyle: this.props.headerStyle
            },
            {
                dataField: 'mostRecentDeadletter',
                width: 23
            },
            {
                dataField: '',
                text: ' ',
                width: 8,
                formatter: (cell, row, rowIndex) => {
                    return (
                        <EditSubscriptionButton subscriptionName={row.name} parentTopic={row.topicName} />
                    );
                },
                headerStyle: this.props.headerStyle
            }
        ];

        let selectRow = {
            bgColor: palerBlue,
            selected: this.props.currentlySelectedName
        };

        return (
            <DataTable
                name='subscriptionList'
                colProps={colProps}
                dataToDisplay={subscriptionArray}
                uniqueKeyColumn='name'
                defaultHover
                onRowClick={this.props.clickFunction}
                selectRow={selectRow}
                searchable
                pagination={{ defaultPageSize: 25, sizePerPageList: [10, 25, 50, 100] }}
            />
        );
    }
}
