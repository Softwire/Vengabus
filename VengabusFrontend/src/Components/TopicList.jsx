import React, { Component } from 'react';
import { DataTable } from './DataTable';
import { palerBlue } from '../colourScheme';

export class TopicList extends Component {

    render() {

        const topics = this.props.topicData;

        const colProps = [
            {
                dataField: 'name',
                text: 'Topic Name',
                width: 34,
                headerStyle: { textAlign: 'center' }
            },
            {
                dataField: 'subscriptionCount',
                text: 'number of subscriptions',
                width: 33,
                headerStyle: { textAlign: 'center' }
            },
            {
                dataField: 'topicStatus',
                text: 'Status',
                width: 33,
                headerStyle: { textAlign: 'center' }
            }
        ];

        let selectRow = {
            bgColor: palerBlue,
            selected: this.props.currentlySelectedName
        };

        return (
            <DataTable
                name='TopicList'
                colProps={colProps}
                keyColumn='name'
                dataToDisplay={topics}
                defaultHover
                onRowClick={this.props.clickFunction}
                selectRow={selectRow}
            />
        );
    }
}
