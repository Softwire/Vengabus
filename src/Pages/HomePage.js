import React, { Component } from 'react';
import { ServiceBusConfigForm } from '../Azure/ServiceBusConfigForm';
import { ExampleServiceBusCall } from '../Azure/ExampleServiceBusCall';

export class HomePage extends Component {
    constructor(props) {
        super(props);

        this.state = {
            connectionString: '',
            queueName: ''
        };
    }

    updateStateOnInput = (e, targetProperty) => {
        var mutation = {};
        mutation[e.target.name] = e.target.value;
        this.setState(mutation);
    };

    render() {
        return (
            <div>
                <ServiceBusConfigForm formCredentials={this.state} updateFormStateOnInput={this.updateStateOnInput} />
                <ExampleServiceBusCall serviceBusConfig={this.state} />
            </div>
        );
    }
}
