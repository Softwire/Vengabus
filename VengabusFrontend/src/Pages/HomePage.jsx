import React, { Component } from 'react';
import { QueueList } from '../Components/QueueList';
import { MessageList } from '../Components/MessageList';
import { css } from 'react-emotion';
import { serviceBusConnection } from '../AzureWrappers/ServiceBusConnection';
import { Glyphicon, Button } from 'react-bootstrap';
import { PAGES, pageSwitcher } from './PageSwitcherService';

export class HomePage extends Component {
    constructor(props) {
        super(props);

        this.state = {
            retrievedData: undefined,
            messageData: [
                { messageId: 10, messageBody: "<a><shipto><name>Ola Nordmann</name><address>Langgt 23</address><city>4000 Stavanger</city><country>Norway</country></shipto><shiporder><item><title>Empire Burlesque</title><note>Special Edition</note><quantity>1</quantity><price>10.90</price></item><item><title>Hide your heart</title><quantity>1</quantity><price>9.90</price></item></shiporder></a>" ,type : "XML"},
                { messageId: 11, messageBody: "<a><b>sdfds</b><c>sdgsdg</c></a>" },
                { messageId: 12, messageBody: "<a>{s}<hipto><name>Ola Nordmann</name><address>Langgt 23</address><city>4000 Stavanger</city><country>Norway</country></shipto><shiporder><item><title>Empire Burlesque</title><note>Special Edition</note><quantity>1</quantity><price>10.90</price></item><item><title>Hide your heart</title><quantity>1</quantity><price>9.90</price></item></shiporder></a>",type: "XML" },
                { messageId: 13, messageBody: "<ashipto><name>Ola Nordmann</name><address>Langgt 23</address><city>4000 Stavanger</city><country>Norway</country></shipto><shiporder><item><title>Empire Burlesque</title><note>Special Edition</note><quantity>1</quantity><price>10.90</price></item><item><title>Hide your heart</title><quantity>1</quantity><price>9.90</price></item></shiporder></a>", type: "XML"},
                { messageId: 14, messageBody: "<shipto><name>Ola Nordmann</name><address>Langgt 23</address><city>4000 Stavanger</city><country>Norway</country></shipto><shiporder><item><title>Empire Burlesque</title><note>Special Edition</note><quantity>1</quantity><price>10.90</price></item><item><title>Hide your heart</title><quantity>1</quantity><price>9.90</price></item></shiporder>",type: "XML" },
                { messageId: 15, messageBody: `{"result":true , "count":42}` ,},
                { messageId: 15, messageBody: `apple` ,},
            ]
        };
    }

    updateRetrievedData = (data) => {
        this.setState({ retrievedData: data });
    };


    render() {
        const queueDivStyle = css`
            width: 30%;
            margin: 10px;
            display: inline-block; /*to allow tables to be displayed side by side*/
        `;
        // Hard coded message to replay for now
        const exampleMessage = {
            MessageProperties: {
                userDefinedProp1: 'value1',
                userDefinedProp2: 'value2'
            },
            MessageBody: 'Hello world!',
            MessageId: 'Message1',
            ContentType: 'null'
        };

        //qq remove hardcoded endpoint names later
        const queueName = 'demoqueue1';
        const topicName = 'demotopic1';
        const subscriptionName = 'demosubscription1';
        const serviceBusService = serviceBusConnection.getServiceBusService();

        return (
            < div >
                <div className={queueDivStyle}>
                    <button onClick={() => pageSwitcher.switchToPage(PAGES.SendMessagePage, exampleMessage)}>Replay Example Message</button>
                    <QueueList queueData={this.state.retrievedData} />
                    <MessageList messageData={this.state.messageData} />
                    {/*qq delete the text in Button once implemented properly*/}
                    <Button onClick={() => serviceBusService.deleteQueueMessages(queueName)} bsStyle="danger">
                        Delete queue messages&#160;
                        <Glyphicon glyph="trash" />
                    </Button>
                    <Button onClick={() => serviceBusService.deleteTopicMessages(topicName)} bsStyle="danger">
                        Delete topic messages&#160;
                        <Glyphicon glyph="trash" />
                    </Button>
                    <Button onClick={() => serviceBusService.deleteSubscriptionMessages(topicName, subscriptionName)} bsStyle="danger">
                        Delete subcription messages&#160;
                        <Glyphicon glyph="trash" />
                    </Button>

                </div>
            </div >
        );
    }
}
