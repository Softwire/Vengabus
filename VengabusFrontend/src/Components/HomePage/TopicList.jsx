import React, { Component } from 'react';
import { DataTable } from '../DataTable';
import { palerBlue } from '../../colourScheme';
import { Button } from 'react-bootstrap';
import { pageSwitcher, PAGES } from '../Pages/PageSwitcherService';
import { EndpointTypes } from '../Helpers/EndpointTypes';


export class TopicList extends Component {

    render() {

        let topics = undefined;
        if (this.props.topicData) {
            topics = [...this.props.topicData];
            for (let i = 0; i < topics.length; i++) {
                topics[i] = { ...topics[i] };
                topics[i].formatterId = i;
            }
        }

     

        const colProps = [
            {
                dataField: 'name',
                width: 31
                headerStyle: this.props.headerStyle
            },
            {
                dataField: 'subscriptionCount',
                width: 30
                headerStyle: this.props.headerStyle,
                align: 'right'
            },
            {
                dataField: 'topicStatus',
                text: 'Status',
                width: 30
            },
            {
                dataField: 'formatterId',
                text: ' ',
                width: 9,
                formatter: (cell, row, rowIndex) => {
                    return (
                        <Button bsSize='xsmall' onClick={() => pageSwitcher.switchToPage(PAGES.CrudPage, { endpointType: EndpointTypes.TOPIC, selectedEndpoint: row.name })} >
                            edit
                        </Button>
                    );
                }
                headerStyle: this.props.headerStyle
            }
        ];

        let selectRow = {
            bgColor: palerBlue,
            selected: this.props.currentlySelectedName
        };

        return (
            <DataTable
                name='TopicList'
                colProps={colProps}
                uniqueKeyColumn='name'
                dataToDisplay={topics}
                defaultHover
                onRowClick={this.props.clickFunction}
                selectRow={selectRow}
                searchable
            />
        );
    }
}
