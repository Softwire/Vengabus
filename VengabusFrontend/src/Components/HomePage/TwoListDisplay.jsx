import React, { Component } from 'react';
import { serviceBusConnection } from '../../AzureWrappers/ServiceBusConnection';
import { QueueList } from './QueueList';
import { TopicList } from './TopicList';
import { MessageList } from './MessageList';
import { css } from 'react-emotion';
import { Breadcrumb, Button } from 'react-bootstrap';
import { SubscriptionList } from './SubscriptionList';
import { EndpointTypes, typeToTitle } from '../../Helpers/EndpointTypes';
import { sharedSizesAndDimensions } from '../../Helpers/SharedSizesAndDimensions';
import { cancellablePromiseCollection } from '../../Helpers/CancellablePromiseCollection';

const messageCount = 500;

export class TwoListDisplay extends Component {
    constructor(props) {
        super(props);
        this.displayHistory = [];
        this.breadCrumbHistory = [{ name: "Home", type: undefined }];
        this.messageButtonDisabled = false;
        this.promiseCollection = new cancellablePromiseCollection();
        this.state = {
            queueData: undefined,
            topicData: undefined,
            subscriptionData: undefined,
            messageData: undefined,
            rightTableType: EndpointTypes.TOPIC
        };
    }

    componentDidMount() {
        serviceBusConnection.registerForUpdatesPrompts(this.resetInitialStateData);

    }
    componentWillUnmount() {
        serviceBusConnection.deregisterForUpdatesPrompts(this.resetInitialStateData);
        this.promiseCollection.cancelAllPromises();

    }

    handleQueueRowClick = (e, row, rowIndex) => {
        this.breadCrumbHistory = [{ name: "Home", type: undefined }, { name: row.name, type: EndpointTypes.QUEUE }];
        this.messageButtonDisabled = true;
        this.setState({
            messageData: undefined,
            rightTableType: EndpointTypes.MESSAGE
        }, this.updateEndpointMessageData);
    }

    handleTopicRowClick = (e, row, rowIndex) => {
        this.breadCrumbHistory = [{ name: "Home", type: undefined }, { name: row.name, type: EndpointTypes.TOPIC }];
        this.setState({
            subscriptionData: undefined,
            rightTableType: EndpointTypes.SUBSCRIPTION
        }, this.updateTopicSubscriptionData);
    }

    handleSubscriptionRowClick = (e, row, rowIndex) => {
        this.breadCrumbHistory[2] = { name: row.name, type: EndpointTypes.SUBSCRIPTION };
        this.messageButtonDisabled = true;
        this.setState({
            messageData: undefined,
            rightTableType: EndpointTypes.MESSAGE
        }, () => this.updateEndpointMessageData(false));
    }

    handleMessageToggle = (isNewTypeDeadLetter) => {
        this.messageButtonDisabled = true;
        this.setState({
            messageData: undefined,
            rightTableType: isNewTypeDeadLetter ? EndpointTypes.DEADLETTER : EndpointTypes.MESSAGE
        }, () => this.updateEndpointMessageData(isNewTypeDeadLetter));
    }

    updateAllQueueData = () => {
        const serviceBusService = serviceBusConnection.getServiceBusService();
        this.promiseCollection.cancelAllPromises(EndpointTypes.QUEUE);
        const fetchedQueueData = this.promiseCollection.addNewPromise(serviceBusService.listQueues(), EndpointTypes.QUEUE);
        fetchedQueueData.then((result) => {
            this.setState({
                queueData: result.map((item) => ({ ...item, mostRecentDeadLetter: null }))
            });
            let currentQueueData = result;
            for (let i = 0; i < result.length; i++) {
                const getMostRecentDLPromise = this.promiseCollection.addNewPromise(serviceBusService.getMostRecentDeadletter(result[i].name), result[i].name);
                getMostRecentDLPromise.then((timeResult) => {
                    const mostRecentDeadLetterTimestamp = timeResult;
                    currentQueueData[i].mostRecentDeadLetter = mostRecentDeadLetterTimestamp;
                    this.setState({
                        queueData: currentQueueData
                    });
                }).catch((e) => { });
            }
        }).catch((e) => { });
    }

    updateAllTopicData = () => {
        const serviceBusService = serviceBusConnection.getServiceBusService();
        this.promiseCollection.cancelAllPromises(EndpointTypes.TOPIC);
        const fetchedTopicData = this.promiseCollection.addNewPromise(serviceBusService.listTopics(), EndpointTypes.TOPIC);
        fetchedTopicData.then(result => {
            this.setState({
                topicData: result
            });
        }).catch(e => { if (!e.isCanceled) { console.log(e); } });
    }

    updateTopicSubscriptionData = () => {
        const topicName = this.breadCrumbHistory[this.breadCrumbHistory.length - 1].name;
        const serviceBusService = serviceBusConnection.getServiceBusService();
        this.promiseCollection.cancelAllPromises(EndpointTypes.SUBSCRIPTION);
        const fetchedSubscriptionData = this.promiseCollection.addNewPromise(serviceBusService.listSubscriptions(topicName), EndpointTypes.SUBSCRIPTION);
        fetchedSubscriptionData.then(result => {
            this.setState({
                subscriptionData: result
            });
        }).catch(e => { if (!e.isCanceled) { console.log(e); } });
    }


    updateEndpointMessageData = (isMessageDeadletters) => {
        const serviceBusService = serviceBusConnection.getServiceBusService();
        let fetchedMessageData;
        const leftTableType = this.breadCrumbHistory[this.breadCrumbHistory.length - 1].type;
        if (leftTableType === EndpointTypes.QUEUE || typeof leftTableType === 'undefined') {
            const queueName = this.breadCrumbHistory[this.breadCrumbHistory.length - 1].name;
            if (isMessageDeadletters) {
                fetchedMessageData = serviceBusService.listQueueDeadLetterMessages(queueName, messageCount);
            } else {
                fetchedMessageData = serviceBusService.listQueueMessages(queueName, messageCount);
            }
        } else {
            const topicName = this.breadCrumbHistory[this.breadCrumbHistory.length - 2].name;
            const subscriptionName = this.breadCrumbHistory[this.breadCrumbHistory.length - 1].name;
            if (isMessageDeadletters) {
                fetchedMessageData = serviceBusService.listSubscriptionDeadLetterMessages(topicName, subscriptionName, messageCount);
            } else {
                fetchedMessageData = serviceBusService.listSubscriptionMessages(topicName, subscriptionName, messageCount);
            }
        }
        this.promiseCollection.cancelAllPromises(EndpointTypes.MESSAGE);
        const wrappedFetchedMessageData = this.promiseCollection.addNewPromise(fetchedMessageData, EndpointTypes.MESSAGE);
        wrappedFetchedMessageData.then((result) => {
            this.messageButtonDisabled = false;
            this.setState({
                messageData: result
            });
        }).catch(e => { if (!e.isCanceled) { console.log(e); } });
    }

    resetInitialStateData = () => {
        this.updateAllTopicData();
        this.updateAllQueueData();
        this.breadCrumbHistory = [{ name: "Home", type: undefined }];
        this.setState({
            rightTableType: EndpointTypes.TOPIC
        });
    };

    getDeadLetterToggleButtonText = (isDeadLetterMessage) => {
        return isDeadLetterMessage ? "Deadletters" : "Live Messages";
    }

    getList = (isForRightHandList) => {
        let typeOfData;
        const currentLeftTable = this.breadCrumbHistory[this.breadCrumbHistory.length - 1];
        let currentSelection;
        if (isForRightHandList) {
            typeOfData = this.state.rightTableType;
        } else {
            //if there is no history the currentLeftTable will be undefined
            //if its undefined its undefined then its the original state and therefore should be queue
            typeOfData = currentLeftTable.type || EndpointTypes.QUEUE;
            currentSelection = currentLeftTable.name;
        }
        const displayStyle = css`
                    display: inline-block;
                    margin:10px;
                `;

        const deadLetterToggleButtonStyle = css`
                margin-right: 0px;
                float: right;
                margin: 9px;
                width: 120px ;
        `;
        const minHeightOfHeader = {
            "minHeight": "92px",
            "height": "92px"
        };

        switch (typeOfData) {
            case EndpointTypes.QUEUE:
                return (
                    <React.Fragment>
                        <div>
                            <h2 id='title'>{typeToTitle(EndpointTypes.QUEUE)}</h2>
                        </div>
                        <QueueList
                            queueData={this.state.queueData}
                            clickFunction={this.handleQueueRowClick}
                            currentlySelectedName={currentSelection}
                            id='QueueTable'
                            headerStyle={minHeightOfHeader}
                        />
                    </React.Fragment>
                );
            case EndpointTypes.TOPIC:
                return (
                    <React.Fragment>
                        <div>
                            <h2 id='title'>{typeToTitle(EndpointTypes.TOPIC)}</h2>
                        </div>
                        <TopicList
                            topicData={this.state.topicData}
                            clickFunction={this.handleTopicRowClick}
                            currentlySelectedName={currentSelection}
                            id='TopicTable'
                            headerStyle={minHeightOfHeader}
                        />
                    </React.Fragment>
                );

            case EndpointTypes.SUBSCRIPTION:
                return (
                    <React.Fragment>
                        <div>
                            <h2 id='title' >{typeToTitle(EndpointTypes.SUBSCRIPTION)}</h2>
                        </div>
                        <SubscriptionList
                            subscriptionData={this.state.subscriptionData}
                            clickFunction={this.handleSubscriptionRowClick}
                            currentlySelectedName={currentSelection}
                            id='SubscriptionTable'
                            headerStyle={minHeightOfHeader}
                        />
                    </React.Fragment>
                );
            case EndpointTypes.MESSAGE:
            case EndpointTypes.DEADLETTER:
                const isDeadLetterMessage = typeOfData === EndpointTypes.DEADLETTER;
                const lastBreadCrumb = this.breadCrumbHistory[this.breadCrumbHistory.length - 1];
                const penultimateBreadCrumb = this.breadCrumbHistory[this.breadCrumbHistory.length - 2];
                return (
                    <React.Fragment>
                        <div >
                            <h2 className={displayStyle} id='title'>{typeToTitle(typeOfData)}</h2>
                            <Button className={deadLetterToggleButtonStyle} onClick={() => this.handleMessageToggle(!isDeadLetterMessage)} disabled={this.messageButtonDisabled} >
                                {this.getDeadLetterToggleButtonText(!isDeadLetterMessage)}
                            </Button>
                        </div>
                        <MessageList
                            id='MessageTable'
                            messageData={this.state.messageData}
                            messageType={typeOfData}
                            endpointType={lastBreadCrumb.type}
                            endpointName={lastBreadCrumb.name}
                            endpointParent={penultimateBreadCrumb.name}
                            headerStyle={minHeightOfHeader}
                            refreshMessageTableHandler={() => {
                                this.setState({
                                    messageData: undefined
                                }, () => this.updateEndpointMessageData(isDeadLetterMessage));

                            }}
                        />
                    </React.Fragment>

                );
            default:
                throw new Error('Invalid endpoint type.');
        }
    }

    HandleBreadCrumbClick = (type, newLength) => {
        switch (type) {
            case (EndpointTypes.QUEUE):
                this.breadCrumbHistory = this.breadCrumbHistory.slice(0, newLength + 1);
                this.updateAllQueueData();
                this.updateEndpointMessageData();
                this.setState({
                    rightTableType: EndpointTypes.MESSAGE
                });
                break;
            case (EndpointTypes.TOPIC):
                this.breadCrumbHistory = this.breadCrumbHistory.slice(0, newLength + 1);
                this.updateAllTopicData();
                this.updateTopicSubscriptionData();
                this.setState({
                    rightTableType: EndpointTypes.SUBSCRIPTION
                });
                break;
            case (EndpointTypes.SUBSCRIPTION):
                this.breadCrumbHistory = this.breadCrumbHistory.slice(0, newLength + 1);
                this.updateTopicSubscriptionData();
                this.updateEndpointMessageData();
                this.setState({
                    rightTableType: EndpointTypes.MESSAGE
                });
                break;
            default:
                this.breadCrumbHistory = this.breadCrumbHistory.slice(0, newLength + 1);
                this.resetInitialStateData();
                break;
        }

    }


    getBreadcrumbElement = () => {
        const breadCrumbStyle = css`
        &.breadcrumb{
          float: left;
            height:${sharedSizesAndDimensions.BREADCRUMBS_HEIGHT}px;
            margin-bottom: 1px;
        }
        `;

        const breadcrumbItems = this.breadCrumbHistory.map((breadCrumb, i) => {
            return (<Breadcrumb.Item id={this.breadCrumbHistory[i].name} onClick={() => this.HandleBreadCrumbClick(breadCrumb.type, i)} active={(i === this.breadCrumbHistory.length - 1)} key={i}>
                {this.breadCrumbHistory[i].name}
            </Breadcrumb.Item>);
        });
        const areOnHomePage = (this.breadCrumbHistory.length === 1);

        if (!areOnHomePage) {
            breadcrumbItems.push(
                (<Breadcrumb.Item key={breadcrumbItems.length} active>
                    {typeToTitle(this.state.rightTableType)}
                </Breadcrumb.Item>)
            );
        }

        return (
            <Breadcrumb className={breadCrumbStyle}>
                {breadcrumbItems}
            </Breadcrumb>
        );
    }

    render() {
        const tableMarginPixels = 15;

        const displayStyle = css`
            width: calc(50% - ${tableMarginPixels * 2}px);
            margin-left:${tableMarginPixels}px;
            margin-right:${tableMarginPixels}px;
            display: inline-block; /*to allow tables to be displayed side by side*/
        `;

        const outerDivDisplay = css`
            display: block;
            `;

        const breadCrumbDisplay = css`
            display: block;
            height:${sharedSizesAndDimensions.BREADCRUMBS_HEIGHT}px;
            margin:2px;
            `;

        const minWidth = css`
            min-width:1000px;
        `;

        const areOnHomePage = (this.breadCrumbHistory.length === 1);
        const totalDiff = sharedSizesAndDimensions.DEFAULT_HEADER_HEIGHT + sharedSizesAndDimensions.BREADCRUMBS_HEIGHT + sharedSizesAndDimensions.BOTTOM_GUTTERING;

        const line = css`
            border-left: 1px solid black;
            display : ${(areOnHomePage) ? "inline-block" : "none"};
            height : calc(100% - ${totalDiff}px); 
            position: fixed;
            z-index: -1;
        `;

        const leftBox = this.getList();
        const rightBox = this.getList(true);



        return (
            <div className={minWidth} >
                <div className={breadCrumbDisplay} >
                    {this.getBreadcrumbElement()}
                </div>
                <div className={outerDivDisplay}>
                    <div className={displayStyle} id="left">
                        {leftBox}
                    </div>
                    <div className={line} />
                    <div className={displayStyle} id="right">
                        {rightBox}
                    </div>
                </div>
            </div>
        );
    }
}

