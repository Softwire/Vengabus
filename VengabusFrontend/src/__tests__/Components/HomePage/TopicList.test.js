import { TopicList } from "../../../Components/HomePage/TopicList";
import renderer from 'react-test-renderer';
import React from 'react';

describe('TopicList', () => {

    it('renders correctly with given props', () => {
        const topics = [
            { name: "topic 1", subscriptionCount: "10", topicStatus: "okay" },
            { name: "topic 2", subscriptionCount: "12", topicStatus: "okay" }
        ];
        let topicsList = renderer.create(
            <TopicList
                topicData={topics}
            />);
        expect(topicsList.toJSON()).toMatchSnapshot();
    });

});