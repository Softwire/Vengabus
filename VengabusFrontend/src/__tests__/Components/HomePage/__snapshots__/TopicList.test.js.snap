// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicList renders correctly with given props 1`] = `
Array [
  <div
    className="form-group"
>
    <input
        className="form-control"
        id="searchBar"
        onChange={[Function]}
        placeholder="Search"
        type="input"
        value=""
    />
</div>,
  .emotion-1 {
  cursor: pointer;
}

.emotion-4 table.table :not(thead) tr:hover {
  border: 1px solid #99ccff;
  background-color: #ecf2ff;
}

<div
    className="react-bootstrap-table emotion-4"
>
    <table
        className="table table-striped table-bordered"
        id="Data"
    >
        <thead>
            <tr
                className={undefined}
            >
                <th
                    style={
                        Object {
                            "width": "31%",
                          }
                    }
                >
                    Name
                </th>
                <th
                    style={
                        Object {
                            "width": "30%",
                          }
                    }
                >
                    Subscription Count
                </th>
                <th
                    style={
                        Object {
                            "width": "30%",
                          }
                    }
                >
                    Status
                </th>
                <th
                    style={
                        Object {
                            "width": "9%",
                          }
                    }
                >
                     
                </th>
            </tr>
        </thead>
        <tbody>
            <tr
                className="emotion-0 emotion-1"
                onClick={[Function]}
                style={Object {}}
            >
                <td>
                    topic 1
                </td>
                <td
                    onClick={[Function]}
                    style={
                        Object {
                            "textAlign": "right",
                          }
                    }
                >
                    10
                </td>
                <td
                    onClick={[Function]}
                >
                    okay
                </td>
                <td
                    onClick={[Function]}
                >
                    <button
                        aria-label="Edit"
                        className="btn btn-xs btn-default"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                    >
                        <span
                            className="glyphicon glyphicon-pencil"
                        />
                    </button>
                </td>
            </tr>
            <tr
                className="emotion-0 emotion-1"
                onClick={[Function]}
                style={Object {}}
            >
                <td>
                    topic 2
                </td>
                <td
                    onClick={[Function]}
                    style={
                        Object {
                            "textAlign": "right",
                          }
                    }
                >
                    12
                </td>
                <td
                    onClick={[Function]}
                >
                    okay
                </td>
                <td
                    onClick={[Function]}
                >
                    <button
                        aria-label="Edit"
                        className="btn btn-xs btn-default"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                    >
                        <span
                            className="glyphicon glyphicon-pencil"
                        />
                    </button>
                </td>
            </tr>
        </tbody>
    </table>
</div>,
  <div
    className="row react-bootstrap-table-pagination"
>
    <div
        className="col-md-6 col-xs-6 col-sm-6 col-lg-6"
    >
        <span
            className="react-bs-table-sizePerPage-dropdown dropdown"
            style={
                Object {
                    "visibility": "visible",
                  }
            }
        >
            <button
                aria-expanded={false}
                className="btn btn-default btn-secondary dropdown-toggle"
                data-toggle="dropdown"
                id="pageDropDown"
                onBlur={[Function]}
                onClick={[Function]}
            >
                25
                 
                <span>
                    <span
                        className="caret"
                    />
                </span>
            </button>
            <ul
                aria-labelledby="pageDropDown"
                className="dropdown-menu "
                role="menu"
            >
                <li
                    className="dropdown-item"
                    role="presentation"
                >
                    <a
                        data-page={10}
                        href="#"
                        onMouseDown={[Function]}
                        role="menuitem"
                        tabIndex="-1"
                    >
                        10
                    </a>
                </li>
                <li
                    className="dropdown-item"
                    role="presentation"
                >
                    <a
                        data-page={25}
                        href="#"
                        onMouseDown={[Function]}
                        role="menuitem"
                        tabIndex="-1"
                    >
                        25
                    </a>
                </li>
                <li
                    className="dropdown-item"
                    role="presentation"
                >
                    <a
                        data-page={50}
                        href="#"
                        onMouseDown={[Function]}
                        role="menuitem"
                        tabIndex="-1"
                    >
                        50
                    </a>
                </li>
                <li
                    className="dropdown-item"
                    role="presentation"
                >
                    <a
                        data-page={100}
                        href="#"
                        onMouseDown={[Function]}
                        role="menuitem"
                        tabIndex="-1"
                    >
                        100
                    </a>
                </li>
            </ul>
        </span>
    </div>
    <div
        className="react-bootstrap-table-pagination-list col-md-6 col-xs-6 col-sm-6 col-lg-6"
    >
        <ul
            className="pagination react-bootstrap-table-page-btns-ul"
        >
            <li
                className="active page-item"
                title="1"
            >
                <a
                    className="page-link"
                    href="#"
                    onClick={[Function]}
                >
                    1
                </a>
            </li>
        </ul>
    </div>
</div>,
]
`;
